### PR TITLE
Jetty 9.4.x 4331 async close complete3

### PR DESCRIPTION
**Pull request by gregw**
This replaces #4398 as a fix for #4331 

This PR breaks out the HttpOutput state into api state and close state.

### CODE REVIEW

**Comment by gregw** [comment]
@lachlan-roberts @sbordet This is ready for proper review.  Would be good to get a green tick early Monday so we can start a release cycle.

File: jetty-server/src/main/java/org/eclipse/jetty/server/HttpOutput.java [hunk 1]
```diff
@@ -63,28 +60,82 @@
...
+       \  v                             v                          v                 |  |
+        +-OPEN/UNREADY----close--->CLOSE/UNREADY----owc----->CLOSING/UNREADY--owcL---+  |
+                      \                         \                                       |
+                       +------------------------->--owcL--------------------------------+
+
+      swl = setWriteListener
+      w = write
+      owc = onWriteComplete(false,null)
+      owcL = onWriteComplete(true,null)
+      isf = isReady()==false
```
Comment (lachlan-roberts) [inline jetty-server/src/main/java/org/eclipse/jetty/server/HttpOutput.java]: This is different to what is on the diagram. This should be irf and irt?

Comment (gregw) [reply jetty-server/src/main/java/org/eclipse/jetty/server/HttpOutput.java]: yes - fixed

**Comment by gregw** [comment]
@sbordet @lachlan-roberts I'm am getting the failure in the attached log very occassionally... only one one laptop... only once in about 1000 runs... but sometimes 10 times in a row!!!
I have no idea what is going on at the moment, so would definitely appreciate a debugging duck at some stage.   But can either of you reproduce on your machines on this branch?
[DelayedServerTest_testPipeline.txt](https://github.com/eclipse/jetty.project/files/3966298/DelayedServerTest_testPipeline.txt)

**Comment by gregw** [comment]
@sbordet @lachlan-roberts I think I found the problem... I was calling the closedCallback and then releasing the buffer.  I've inverted that order and now it does not fail.

**Comment by gregw** [comment]
92% code coverage of HttpOutput with current tests

**Comment by gregw** [comment]
@sbordet review please

**Review by sbordet — APPROVED** [review]

**Pull request merged**